In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/student-performance-data/student_data.csv


In [2]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np

import plotly.express as px
import plotly.graph_objs as go
from plotly.subplots import make_subplots
import seaborn as sns

pd.set_option('display.max_columns', 60)
pd.set_option('display.max_rows', 60)

In [3]:
df = pd.read_csv('../input/student-performance-data/student_data.csv')

# Exploratory Data Analysis

In [4]:
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

In [5]:
df['G_final'] = ((df['G1'] + df['G2'] + df['G3']) / 3).round(2)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   school      395 non-null    object 
 1   sex         395 non-null    object 
 2   age         395 non-null    int64  
 3   address     395 non-null    object 
 4   famsize     395 non-null    object 
 5   Pstatus     395 non-null    object 
 6   Medu        395 non-null    int64  
 7   Fedu        395 non-null    int64  
 8   Mjob        395 non-null    object 
 9   Fjob        395 non-null    object 
 10  reason      395 non-null    object 
 11  guardian    395 non-null    object 
 12  traveltime  395 non-null    int64  
 13  studytime   395 non-null    int64  
 14  failures    395 non-null    int64  
 15  schoolsup   395 non-null    object 
 16  famsup      395 non-null    object 
 17  paid        395 non-null    object 
 18  activities  395 non-null    object 
 19  nursery     395 non-null    o

Perguntas:

1. Qual a média das notas G1, G2 e G3 dos alunos?
2. Estudantes com maior free time possuem em média notas mais altas?
3. O gênero do estudante impacta em seu desempenho?
4. Os estudantes possuem em média maior desempenho acadêmico sob a guarda paterna ou materna?
5. Como as faltas dos estudantes impactam na média das 3 notas?
6. Quais são os tipos de empregos predominantes entre os pais dos alunos?

**1. Qual a média das notas G1, G2 e G3 dos alunos?**

In [7]:
df['G1'].mean()

10.90886075949367

In [8]:
df['G2'].mean()

10.713924050632912

In [9]:
df['G3'].mean()

10.415189873417722

As médias de notas dos alunos são 10.9, 10.7 e 10.4, respectivamente.

**2. Quais as quantidades de free time ideais para alunos?**

Para determinar isto, será feita uma análise de quais quantidades de free time há notas em média, maiores que a média das notas G1, G2 e G3.

In [10]:
df1 = df.groupby(by = 'freetime').agg({'G_final':'mean', 'G1':'mean', 'G2':'mean', 'G3':'mean'}).reset_index()
df1.sort_values(by = 'G_final')

freetime    G_final         G1         G2         G3
0         1  10.141053  10.157895  10.421053   9.842105
2         3  10.144268  10.369427  10.280255   9.783439
3         4  10.608435  10.834783  10.565217  10.426087
4         5  11.475000  11.775000  11.350000  11.300000
1         2  11.780625  12.046875  11.734375  11.562500

In [11]:
fig = px.bar(data_frame = df1,
             x = 'freetime', y = 'G_final',
             text_auto = '.3s', template = 'plotly_dark')

fig.update_layout(
    title = {
        'text': "Média de notas por tempo livre"},
    xaxis_title = 'Tempo Livre',
    yaxis_title = 'Média de Notas',
    width = 1000,
    height = 500,
    font_family = "Arial",
    font_color = "White",
    title_font_family = "Arial",
    title_font_color = "White")

fig.show()

Os estudantes que possuem 2 e 5 horas de free time, possuem notas acima da média.

**3. O gênero do estudante impacta em seu desempenho?**

In [12]:
df2 = df.groupby(by = 'sex').agg({'G1':'mean',
                            'G2':'mean',
                            'G3':'mean',
                            'G_final':'mean'}).reset_index().sort_values(by = ['G_final', 'G1', 'G2', 'G3'],
                                                                        ascending = False)
df2

sex         G1         G2         G3    G_final
1   M  11.229947  11.074866  10.914439  11.073048
0   F  10.620192  10.389423   9.966346  10.325000

In [13]:
fig = make_subplots(rows = 2, cols = 2)

fig = make_subplots(rows = 2,
                    cols = 2,
                    subplot_titles = [
                        'G1',  # 1. título do subplot 
                        'G2',# 2. título do subplot
                        'G3',   # 3. título do subplot
                        'G médio', #  4. título do subplot
])

# Adiciona o primeiro grafico e especifica a linha e a coluna
fig.add_trace(go.Bar(x = df2['sex'], 
                     y = df2['G1'], textposition='auto',
                     name = 'G1'), row = 1, col = 1)

# adiciona o segundo gráfico
fig.add_trace(go.Bar(x = df2['sex'], 
                     y = df2['G2'], textposition='auto',
                     name = 'G2'), row = 1, col = 2)

# adiciona o terceiro gráfico
fig.add_trace(go.Bar(x = df2['sex'], 
                     y = df2['G3'], textposition='auto',
                     name = 'G3'), row = 2, col = 1)

# adiciona o quarto gráfico
fig.add_trace(go.Bar(x = df2['sex'], 
                     y = df2['G_final'], textposition='auto',
                     name = 'G Médio'), row = 2, col = 2)

# Update xaxis propriedades
fig.update_xaxes(title_text = "Sexos", row=1, col=1)
fig.update_xaxes(title_text = "Sexos", row=1, col=2)
fig.update_xaxes(title_text = "Sexos", row=2, col=1)
fig.update_xaxes(title_text = "Sexos", row=2, col=2)

# Update yaxis propriedades
fig.update_yaxes(title_text = "Nota G1", row=1, col=1)
fig.update_yaxes(title_text = "Nota G2", row=1, col=2)
fig.update_yaxes(title_text = "Nota G3", row=2, col=1)
fig.update_yaxes(title_text = "Nota G Médio", row=2, col=2)

fig.update_layout(title = 'Notas por Sexos', title_x = 0.5)

Os estudantes do sexo masculino possuem maiores notas, em média.

**4. Os estudantes possuem em média maior desempenho acadêmico sob a guarda paterna ou materna?**

In [14]:
df3 = df.groupby(by = 'guardian').agg({'G1':'mean',
                            'G2':'mean',
                            'G3':'mean',
                            'G_final':'mean'}).reset_index().sort_values(by = ['G_final', 'G1', 'G2', 'G3'],
                                                                        ascending = False)
df3

guardian         G1         G2         G3    G_final
0   father  11.111111  11.155556  10.688889  10.985111
1   mother  10.882784  10.677656  10.483516  10.681062
2    other  10.562500   9.781250   9.062500   9.802188

Os estudantes sob a guarda do pai possuem as maiores notas, enquanto aqueles que estão sob a guarda de indivíduos que não são os seus pais ou mães, apresentam menor desempenho.

**5. Como a falta dos estudantes impactam na média das 3 notas?**

In [15]:
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  G_final  
0     1       3         6   5   6   6     5.67  
1     1       3         4   5   5   6     5.33  
2     3       3        10   7   8  10     8.33  
3     1       5         2  15  14  15    14.67  
4     2       5         4   6  10  10     8.67

In [16]:
df4 = df.groupby(by = 'absences').agg({'G1':'mean',
                            'G2':'mean',
                            'G3':'mean',
                            'G_final':'mean'}).reset_index().sort_values(by = ['absences', 'G_final', 'G1', 'G2', 'G3'],
                                                                        ascending = False)
df4

absences         G1         G2         G3    G_final
33        75  10.000000   9.000000   9.000000   9.330000
32        56   9.000000   9.000000   8.000000   8.670000
31        54  11.000000  12.000000  11.000000  11.330000
30        40  13.000000  11.000000  11.000000  11.670000
29        38   8.000000   9.000000   8.000000   8.330000
28        30   8.000000   8.000000   8.000000   8.000000
27        28  10.000000   9.000000   9.000000   9.330000
26        26   7.000000   6.000000   6.000000   6.330000
25        25   7.000000  10.000000  11.000000   9.330000
24        24  18.000000  18.000000  18.000000  18.000000
23        23  13.000000  13.000000  13.000000  13.000000
22        22   9.333333   8.333333   8.000000   8.553333
21        21  17.000000  18.000000  18.000000  17.670000
20        20  12.750000  11.250000  11.500000  11.832500
19        19  11.000000   9.000000  10.000000  10.000000
18        18  10.000000   8.000000   8.400000   8.800000
17        17  10.000000  10.000000  10.000000  10.000000
16        16  10.142857  10.000000   9.857143   9.998571
15        15   8.000000   8.000000   8.333333   8.110000
14        14  10.000000   9.500000   9.333333   9.610833
13        13  11.666667  11.666667  12.666667  12.000000
12        12  10.666667  10.916667  10.750000  10.777500
11        11  12.333333  11.666667  11.666667  11.886667
10        10  11.176471  10.764706  11.235294  11.058235
9          9  13.333333  12.666667  13.333333  13.110000
8          8  10.863636  10.454545  10.590909  10.636364
7          7  11.428571  11.285714  11.857143  11.522857
6          6  10.838710  11.419355  11.774194  11.344194
5          5  11.200000  11.000000  11.600000  11.268000
4          4  10.830189  11.113208  11.283019  11.075094
3          3  11.500000  11.500000  11.500000  11.500000
2          2  11.630769  11.800000  11.907692  11.779385
1          1  12.333333  12.666667  13.000000  12.666667
0          0  10.547826   9.886957   8.443478   9.626174

**5.1 Quantas faltas os estudantes de maior desempenho possuem em média?**

In [17]:
df4[df4['G_final'] > df4['G_final'].mean()].sort_values(by = 'G_final',
                                                       ascending = False)

absences         G1         G2         G3    G_final
24        24  18.000000  18.000000  18.000000  18.000000
21        21  17.000000  18.000000  18.000000  17.670000
9          9  13.333333  12.666667  13.333333  13.110000
23        23  13.000000  13.000000  13.000000  13.000000
1          1  12.333333  12.666667  13.000000  12.666667
13        13  11.666667  11.666667  12.666667  12.000000
11        11  12.333333  11.666667  11.666667  11.886667
20        20  12.750000  11.250000  11.500000  11.832500
2          2  11.630769  11.800000  11.907692  11.779385
30        40  13.000000  11.000000  11.000000  11.670000
7          7  11.428571  11.285714  11.857143  11.522857
3          3  11.500000  11.500000  11.500000  11.500000
6          6  10.838710  11.419355  11.774194  11.344194
31        54  11.000000  12.000000  11.000000  11.330000
5          5  11.200000  11.000000  11.600000  11.268000
4          4  10.830189  11.113208  11.283019  11.075094
10        10  11.176471  10.764706  11.235294  11.058235

**6. Quais são os tipos de empregos predominantes entre os pais dos alunos?**

In [18]:
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  G_final  
0     1       3         6   5   6   6     5.67  
1     1       3         4   5   5   6     5.33  
2     3       3        10   7   8  10     8.33  
3     1       5         2  15  14  15    14.67  
4     2       5         4   6  10  10     8.67

In [19]:
df['count'] = 1

In [20]:
df.groupby(by = ['Fjob']).agg({'count':'count'}).reset_index().sort_values(by = 'count',
                                                                          ascending = False)

Fjob  count
2     other    217
3  services    111
4   teacher     29
0   at_home     20
1    health     18

Os tipos de trabalho predominantes dos pais das crianças estão na categoria services e professor, enquanto que as mães também se enquadram em services, mas realizam trabalhos domésticos como segunda maior colocação. Ambos os sexos possuem trabalhos diversos.

In [21]:
df.groupby(by = ['Mjob']).agg({'count':'count'}).reset_index().sort_values(by = 'count',
                                                                          ascending = False)

Mjob  count
2     other    141
3  services    103
0   at_home     59
4   teacher     58
1    health     34